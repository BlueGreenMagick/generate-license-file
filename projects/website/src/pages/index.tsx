import styled from "@emotion/styled";
import Layout from "@theme/Layout";
import React from "react";
import { MonacoEditor } from "../components/MonacoEditor/MonacoEditor";
import { Tab } from "../components/Tabs/Tab";
import { Tabs } from "../components/Tabs/Tabs";
import { Terminal } from "../components/Terminal";

export default function Home(): JSX.Element {
  return (
    <Layout
      title={`Generate License File | Docs`}
      description="Docs website for the generate-license-file npm package."
    >
      <header>
        <Hero>
          <Intro>
            <Icon />
            <Title>Generate License File</Title>
            <Subtitle>
              Generate a text file containing all of the licencss for your <br /> production,
              third-party dependencies.
            </Subtitle>
          </Intro>
          <Examples>
            <Tabs>
              <Tab label="CLI">
                <Terminal />
              </Tab>
              <Tab label="Library">
                <MonacoEditor />
              </Tab>
            </Tabs>
          </Examples>
        </Hero>
      </header>
    </Layout>
  );
}

const Hero = styled.div`
  padding: 80px 70px 100px 70px;
  min-height: 100vh;
`;

const Icon = styled.div`
  background-image: url("/img/glf-icon-3.svg");
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  height: 100px;
  width: 100%;
  margin-bottom: 20px;
`;

const Title = styled.h1`
  font-size: 3rem;
  color: var(--ifm-heading-color);
  font-family: var(--ifm-heading-font-family);
  font-weight: var(--ifm-heading-font-weight);
  line-height: var(--ifm-heading-line-height);
  margin: var(--ifm-heading-margin-top) 0 30px 0;
`;

const Subtitle = styled.p`
  font-size: 1.5rem;
  max-width: 700px;
  margin-bottom: 64px;
`;

const Intro = styled.div`
  text-align: center;
  display: flex;
  flex-direction: column;
  align-items: center;
  position: relative;
  max-width: 900px;
  width: 100%;
  margin: 0 auto;
`;

const Examples = styled.div`
  text-align: center;
  margin: auto;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;
