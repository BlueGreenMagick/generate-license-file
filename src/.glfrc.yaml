inputs:
  - ./package.json
replace:
  exponential-backoff@3.1.1: ./nx.json
