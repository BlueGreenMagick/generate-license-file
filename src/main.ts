import { promisify } from "util";
import glob from "glob";
import { readFileAsync, doesFileExist } from "./utils/file.utils";
import { ILicense } from "./models/license.interface";

const globAsync: (pattern: string, options?: glob.IOptions) => Promise<string[]> = promisify(glob);
const UTF8: string = "utf-8";

export async function getProjectLicenses(path: string): Promise<ILicense[]> {

  try {
    const dependencyLicenses: Map<string, ILicense> = new Map<string, ILicense>();

    if (!await doesFileExist(path)) {
      throw new Error("Cannot find file " + path);
    }

    const file: string = await readFileAsync(path, { encoding: UTF8 });
    const packageJson: any = JSON.parse(file);

    const dependencies: string[] = packageJson.dependencies ?? [];

    for (const dep in dependencies) {
      if (dependencies.hasOwnProperty(dep)) {
        const licenseFiles: string[] = await findLicenses(dep);
          for(const license of licenseFiles) {
            const content: string = await readFileAsync(license, { encoding: UTF8 });

            if (!dependencyLicenses.has(content)) {
              dependencyLicenses.set(content, {
                dependencies: [],
                content
              });
            }

            const foundLicence: ILicense | undefined = dependencyLicenses.get(content);
            foundLicence?.dependencies.push(dep);
          }
        }
    }

    return Array.from(dependencyLicenses.values());
  } catch (error) {
    console.error(error);
    return Promise.reject();
  }
}

async function findLicenses(packageName: string): Promise<string[]> {
  return globAsync(`node_modules/${packageName}/**/LICEN{S,C}E*`, { nocase: true });
}

(async () => {
	const licenses: ILicense[] = await getProjectLicenses("./package.json");
})();
