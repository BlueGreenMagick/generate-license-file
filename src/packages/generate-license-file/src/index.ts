/* istanbul ignore file */
export { generateLicenseFile } from "./lib/generateLicenseFile";
export { getLicenseFileText } from "./lib/getLicenseFileText";
export { getProjectLicenses } from "./lib/getProjectLicenses";
export { LineEnding } from "./lib/lineEndings";
export { ILicense } from "./lib/models/license";
