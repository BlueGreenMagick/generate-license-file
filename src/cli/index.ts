import arg from "arg";
import { pong } from "cli-spinners";
import { prompt } from "enquirer";
import ora, { Ora } from "ora";
import path from "path";
import { generateLicenseFile } from "../generateLicenseFile";
import { doesFileExist } from "../utils/file.utils";
import { IArguments } from "./arguments.interface";
import { validArguments } from "./valid-arguments";

function parseArgumentsIntoOptions(rawArgs: string[]): IArguments {
  const args: arg.Result<any> = arg(validArguments, {
    argv: rawArgs.slice(2)
  });

  return {
    input: args["--input"] || undefined,
    output: args["--output"] || undefined,
    overwriteOutput: args["--overwrite"] || undefined,
    eol: args["--eol"] || undefined
  };
}

async function promptForAnswers(options: IArguments): Promise<IArguments> {
  if (!options.input) {
    const answer: any = await prompt({
      type: "input",
      name: "input",
      initial: (await doesFileExist("./package.json")) ? "./package.json" : "",
      message: "package.json location:"
    });

    options.input = answer.input;
  }

  while (!options.output || options.overwriteOutput === false) {
    const answer: any = await prompt({
      type: "input",
      name: "output",
      initial: "3rd-party-licenses.txt",
      message: "Output file location:"
    });

    options.output = answer.output;

    if (await doesFileExist(options.output)) {
      const yesNoAnswer: any = await prompt({
        type: "confirm",
        name: "overwriteOutput",
        message: "The given output file already exists and will be overwritten. Is this OK?"
      });

      options.overwriteOutput = yesNoAnswer.overwriteOutput;
    }
  }

  return options;
}

export async function cli(args: string[]): Promise<void> {
  let options: IArguments = parseArgumentsIntoOptions(args);
  options = await promptForAnswers(options);

  const spinner: Ora = ora({
    spinner: pong,
    text: "Resolving licenses..."
  });
  spinner.start();

  const directory: string = path.dirname(options.input);

  await generateLicenseFile(directory, options.output, options.eol);

  spinner.stop();
}
