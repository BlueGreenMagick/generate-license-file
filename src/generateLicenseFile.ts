import { getLicenseFileText } from "./getLicenseFileText";
import { writeFileAsync } from "./utils/file.utils";

/**
 * Used to specify which line endings to use in the generated file
 *
 * `windows` = "\r\n"
 *
 * `posix` = "\n"
 */
export type LineEnding = "windows" | "posix";

export function isValidEol(input: string | undefined): input is LineEnding {
  return input === "windows" || input === "posix" || input === undefined;
}

/**
 * Scans the project found at the given path and creates a license file at the given output location
 * @param pathToPackageJson A path to the package.json for the project
 * @param outputPath A file path for the resulting license file
 * @optional @param lineEnding "windows" or "posix". Will use the system default if omitted
 */
export async function generateLicenseFile(
  pathToPackageJson: string,
  outputPath: string,
  lineEnding?: LineEnding
): Promise<void>;

/**
 * Scans the projects found at the given paths and creates a license file at the given output location
 * @param pathsToPackageJsons A path to the package.json for the project
 * @param outputPath A file path for the resulting license file
 * @optional @param lineEnding "windows" or "posix". Will use the system default if omitted
 */
export async function generateLicenseFile(
  pathsToPackageJsons: string[],
  outputPath: string,
  lineEnding?: LineEnding
): Promise<void>;

export async function generateLicenseFile(
  pathsToPackageJsons: string[] | string,
  outputPath: string,
  lineEnding?: LineEnding
): Promise<void> {
  if (typeof pathsToPackageJsons === "string") {
    pathsToPackageJsons = [pathsToPackageJsons];
  }

  const licenseFileText: string = await getLicenseFileText(pathsToPackageJsons, lineEnding);
  await writeFileAsync(outputPath, licenseFileText);
}
