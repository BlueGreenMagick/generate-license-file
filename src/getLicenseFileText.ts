import * as os from "os";
import { LineEnding } from "./generateLicenseFile";
import { getProjectLicensesInternal } from "./internal/getProjectLicensesInternal";
import { License } from "./models/license";

const SUFFIX: string = "-----------";
const FOOTER: string =
  "This file was generated with generate-license-file! https://www.npmjs.com/package/generate-license-file";

/**
 * Scans the project found at the given path and returns a string containing the licenses for all the dependencies
 * @param path A path to a directory containing a package.json
 * @optional @param lineEnding "windows" or "posix". Will use the system default if omitted
 * @returns A promise that resolves to the license file text
 */
export async function getLicenseFileText(path: string, lineEnding?: LineEnding): Promise<string> {
  const EOL = getLineEnding(lineEnding);
  const licenses: License[] = await getProjectLicensesInternal(path);
  let licenseFile = "";

  for (const license of licenses) {
    licenseFile += license.format(EOL) + EOL + EOL + SUFFIX + EOL + EOL;
  }

  licenseFile += FOOTER + EOL;
  return licenseFile;
}

function getLineEnding(lineEndings?: LineEnding): string {
  switch (lineEndings) {
    case "windows":
      return "\r\n";
    case "posix":
      return "\n";
    default:
      return os.EOL;
  }
}
