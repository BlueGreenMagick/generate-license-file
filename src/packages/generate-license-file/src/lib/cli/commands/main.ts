import { Command } from "@commander-js/extra-typings";
import { loadConfigFile } from "../config";
import { Inputs } from "../args/inputs";
import { Output } from "../args/output";
import { Eol } from "../args/eol";
import { LineEnding } from "../../lineEndings";
import { generateLicenseFile } from "../../generateLicenseFile";
import { spinner } from "../spinner";

export type CombinedConfig = {
  inputs?: string[];
  output?: string;
  eol?: string;
  ci?: boolean;
  overwrite?: boolean;
  spinner?: boolean;
};

export interface ProgramOptions {
  inputs: string[];
  output: string;
  eol?: LineEnding;
  showSpinner: boolean;
}

export const mainCommand = new Command()
  .name("generate-license-file")
  .description("// TODO")
  .option("-c,--config <path>", "")
  .option("-i,--input <paths...>", "")
  .option("-o,--output <path>", "")
  .option("--eol <eol>", "")
  .option("--ci", "")
  .option("--no-spinner", "")
  .option("--overwrite", "")
  .action(async givenArgs => {
    const cliArgs = {
      spinner: givenArgs.spinner ? undefined : false,
      ci: givenArgs.ci,
      eol: givenArgs.eol,
      inputs: givenArgs.input,
      output: givenArgs.output,
      overwrite: givenArgs.overwrite,
    };

    const configFile = await loadConfigFile(givenArgs.config);

    const filteredCliArgs = Object.fromEntries(
      Object.entries(cliArgs).filter(([, v]) => v !== undefined),
    );

    const combinedConfig = {
      ...configFile,
      ...filteredCliArgs,
    };

    const { inputs, showSpinner, output, eol } = await parseArgumentsIntoOptions(combinedConfig);

    if (showSpinner) {
      spinner.start();
    }

    await generateLicenseFile(inputs, output, eol);
    spinner.stop();
  });

async function parseArgumentsIntoOptions(config: CombinedConfig): Promise<ProgramOptions> {
  if (config.ci) {
    config.spinner = false;

    try {
      return await getOptionsOrThrow(config);
    } catch (e: unknown) {
      const errorMessage = getMessageFromCaughtUnknown(e);
      throw new Error(`Error parsing arguments in --ci mode: ${errorMessage}`);
    }
  }

  return await promptForMissingOptions(config);
}

async function getOptionsOrThrow(config: CombinedConfig): Promise<ProgramOptions> {
  const inputs = await new Inputs().parse(config);
  const output = await new Output().parse(config);
  const eol = await new Eol().parse(config);
  const showSpinner = config.spinner ?? true;

  return { inputs, output, eol, showSpinner };
}

async function promptForMissingOptions(config: CombinedConfig): Promise<ProgramOptions> {
  const inputs = await new Inputs().resolve(config);
  const output = await new Output().resolve(config);
  const eol = await new Eol().resolve(config);
  const showSpinner = config.spinner ?? true;

  return { inputs, output, eol, showSpinner };
}

function getMessageFromCaughtUnknown(e: unknown): string {
  if (e instanceof Error) {
    return e.message;
  }

  if (typeof e === "string") {
    return e;
  }

  return "Unknown error";
}
