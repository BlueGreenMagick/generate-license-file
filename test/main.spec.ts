import * as main from "../src/main";
import consoleUtils from "../src/utils/console.utils";
import * as fileUtils from "../src/utils/file.utils";
import * as licenceUtils from "../src/utils/licence.utils";
import { Project } from "../src/utils/licence.utils";

class MockStream {
  allText: string = "";
  ended: boolean = false;

  once = jest.fn();
  write = jest.fn(text => {
    this.allText += text;
  });
  end = jest.fn(text => {
    this.allText += text;
    this.ended = true;
  });
}

describe("main", () => {
  let mockDoesFolderExist: jest.SpyInstance;

  let mockStream: MockStream;
  let mockCreateWriteStream: jest.SpyInstance;

  let subject: typeof main;

  beforeEach(() => {
    jest.spyOn(consoleUtils, "error").mockImplementation();

    mockDoesFolderExist = jest.spyOn(fileUtils, "doesFolderExist").mockImplementation();

    mockStream = new MockStream();

    mockCreateWriteStream = jest
      .spyOn(fileUtils, "createWriteStream")
      .mockImplementation()
      .mockReturnValue(mockStream as any);

    subject = require("../src/main");
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  describe("generateLicenseFile", () => {
    it("should write the found licences to the given file name", async () => {
      mockDoesFolderExist.mockReturnValue(Promise.resolve(true));

      const mockProject: Project = {
        lib1: { licenseFile: "./lib1/LICENSE.md" },
        lib2: { licenseFile: "./lib2/LICENSE.md" },
        lib3: { licenseFile: "./lib3/LICENSE.md" }
      };

      jest.spyOn(licenceUtils, "getProject").mockImplementation(() => Promise.resolve(mockProject));
      jest.spyOn(fileUtils, "doesFileExist").mockImplementation(() => Promise.resolve(true));
      jest
        .spyOn(fileUtils, "readFileAsync")
        .mockImplementationOnce(() => Promise.resolve("licence content"))
        .mockImplementationOnce(() => Promise.resolve("different licence content"))
        .mockImplementationOnce(() => Promise.resolve("licence content"));

      await subject.generateLicenseFile("../", "output.txt");

      expect(mockStream.once).toHaveBeenCalledTimes(1);
      const onceCallback = mockStream.once.mock.calls[0][1];
      onceCallback();

      expect(mockStream.allText).toMatchSnapshot();
      expect(mockCreateWriteStream.mock.calls[0][0]).toBe("output.txt");
      expect(mockStream.ended).toBe(true);
    });
  });

  describe("getProjectLicenses", () => {
    it("should throw if the given directory cannot be found", async () => {
      mockDoesFolderExist.mockReturnValue(Promise.resolve(false));

      try {
        await subject.getProjectLicenses("../");
        fail("should have thrown");
      } catch (e) {
        expect(e).toBe(undefined);
      }

      expect(consoleUtils.error).toHaveBeenCalledWith("Cannot find directory ../");
      expect(mockDoesFolderExist).toHaveBeenCalledWith("../");
    });

    it("should read in the license file for a dependency if it exists", async () => {
      mockDoesFolderExist.mockReturnValue(Promise.resolve(true));

      const mockProject: Project = {
        lib1: { licenseFile: "./lib1/LICENSE.md" },
        lib2: { licenseFile: "./lib2/LICENSE.md" },
        lib3: { licenseFile: "./lib3/LICENSE.md" }
      };

      jest.spyOn(licenceUtils, "getProject").mockImplementation(() => Promise.resolve(mockProject));
      jest.spyOn(fileUtils, "doesFileExist").mockImplementation(() => Promise.resolve(true));
      jest.spyOn(fileUtils, "readFileAsync").mockImplementation(() => Promise.resolve(""));

      await subject.getProjectLicenses("../");

      expect(fileUtils.doesFileExist).toHaveBeenCalledWith("./lib1/LICENSE.md");
      expect(fileUtils.doesFileExist).toHaveBeenCalledWith("./lib2/LICENSE.md");
      expect(fileUtils.doesFileExist).toHaveBeenCalledWith("./lib3/LICENSE.md");

      expect(fileUtils.readFileAsync).toHaveBeenCalledWith("./lib1/LICENSE.md", {
        encoding: "utf-8"
      });
      expect(fileUtils.readFileAsync).toHaveBeenCalledWith("./lib2/LICENSE.md", {
        encoding: "utf-8"
      });
      expect(fileUtils.readFileAsync).toHaveBeenCalledWith("./lib3/LICENSE.md", {
        encoding: "utf-8"
      });
    });

    it("should return all dependencies for a licence", async () => {
      mockDoesFolderExist.mockReturnValue(Promise.resolve(true));

      const mockProject: Project = {
        lib1: { licenseFile: "./lib1/LICENSE.md" },
        lib2: { licenseFile: "./lib2/LICENSE.md" },
        lib3: { licenseFile: "./lib3/LICENSE.md" }
      };

      jest.spyOn(licenceUtils, "getProject").mockImplementation(() => Promise.resolve(mockProject));
      jest.spyOn(fileUtils, "doesFileExist").mockImplementation(() => Promise.resolve(true));
      jest
        .spyOn(fileUtils, "readFileAsync")
        .mockImplementation(() => Promise.resolve("licence content"));

      const result = await subject.getProjectLicenses("../");

      expect(result).toEqual([
        { content: "licence content", dependencies: ["lib1", "lib2", "lib3"] }
      ]);
    });

    it("should group dependencies by their licence", async () => {
      mockDoesFolderExist.mockReturnValue(Promise.resolve(true));

      const mockProject: Project = {
        lib1: { licenseFile: "./lib1/LICENSE.md" },
        lib2: { licenseFile: "./lib2/LICENSE.md" },
        lib3: { licenseFile: "./lib3/LICENSE.md" }
      };

      jest.spyOn(licenceUtils, "getProject").mockImplementation(() => Promise.resolve(mockProject));
      jest.spyOn(fileUtils, "doesFileExist").mockImplementation(() => Promise.resolve(true));
      jest
        .spyOn(fileUtils, "readFileAsync")
        .mockImplementationOnce(() => Promise.resolve("licence content"))
        .mockImplementationOnce(() => Promise.resolve("different licence content"))
        .mockImplementationOnce(() => Promise.resolve("licence content"));

      const result = await subject.getProjectLicenses("../");

      expect(result).toEqual([
        { content: "licence content", dependencies: ["lib1", "lib3"] },
        { content: "different licence content", dependencies: ["lib2"] }
      ]);
    });
  });
});
