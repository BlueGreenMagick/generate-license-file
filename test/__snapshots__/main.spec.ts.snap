// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main generateLicenseFile should write the found licences to the given file name 1`] = `
"The following NPM packages may be included in this product:

 - lib1
 - lib3

These packages each contain the following license and notice below:

licence content

-----------

The following NPM package may be included in this product:

 - lib2

This package contains the following license and notice below:

different licence content

-----------

This file was generated with generate-license-file! https://www.npmjs.com/package/generate-license-file"
`;
